### PR TITLE
Add accounts_data_len to bank snapshot

### PR DESCRIPTION
**Pull request by brooksprumo**
#### Problem

The total accounts data size needs to be deterministic across the cluster. When loading from a snapshot, the bank needs to get the `accounts_data_len` from the snapshot, but it currently is not there.

#### Summary of Changes

Add `accounts_data_len` to the bank snapshot

#### Details

* Since the bank snapshot already have an unused u64 field, I'm using that spot to avoid any issues around snapshot versioning.
* The Bank will get an approximate accounts data len on startup via generate_index(). That number is a bit different from the `Bank::accounts_data_len` field. To get more data on this difference from multiple nodes over time, I've added a datapoint into `Bank::new_from_fields()`.
* I'm getting this in early, so that (1) I can get the datapoints from 👆,  and (2) there will be sufficient time to pass so that all valid snapshots will now contain the accounts_data_len field.

Part of #21604

### CODE REVIEW

File: runtime/src/serde_snapshot/newer.rs [hunk 1]
```diff
@@ -39,8 +39,7 @@ struct DeserializableVersionedBank {
     ns_per_slot: u128,
     genesis_creation_time: UnixTimestamp,
     slots_per_year: f64,
-    #[allow(dead_code)]
-    unused: u64,
```
Comment (mvines) [inline runtime/src/serde_snapshot/newer.rs:43]: well that's convenient!  No issue with newer code reading an older snapshot and decoding `accounts_data_len` as 0?

Comment (brooksprumo) [reply runtime/src/serde_snapshot/newer.rs:43]: > No issue with newer code reading an older snapshot and decoding `accounts_data_len` as 0?

Correct, no issue. The main reason is that `Bank::new_from_fields()` doesn't use the snapshot (i.e. `fields`) to set its `accounts_data_len` field yet [1]; that value is calculated during `generate_index()` [2] and passed up as a parameter to `new_from fields()` [3].

So this PR will start _writing_ real-ish values into snapshots, but not _reading_ that value yet. PR #23744 will start reading the value (and I elaborate a bit more on the safety of `fields.accounts_data_len` too).

[1] https://github.com/solana-labs/solana/blob/6b0d34d70dc08beb3d94cc8e30423883de73646b/runtime/src/bank.rs#L2007

[2] https://github.com/solana-labs/solana/blob/6b0d34d70dc08beb3d94cc8e30423883de73646b/runtime/src/serde_snapshot.rs#L349-L357

[3] https://github.com/solana-labs/solana/blob/6b0d34d70dc08beb3d94cc8e30423883de73646b/runtime/src/bank.rs#L1930-L1939

**Review by mvines — APPROVED** [review]
lgtm, beyond that `accounts_data_len` name that I still think is a little misleading.

`total_accounts_size`?

**Comment by brooksprumo** [comment]
> lgtm, beyond that `accounts_data_len` name that I still think is a little misleading.
> 
> `total_accounts_size`?

Opened https://github.com/solana-labs/solana/issues/23759 to track this issue.

**Pull request merged**